File: Snake/Snake.cpp
#include "Snake.h"
Snake::Snake()
{
	setX(8);
	setY(16);
	direction = Direction::RIGHT;
}

Snake::Snake(int x, int y, Direction direction)
{
	setX(x);
	setY(y);
	this->direction = direction;
}

std::string Snake::getSpriteHead()
{
	return spriteHead;
}

std::string Snake::getSpriteBody()
{
	return spriteBody;
}

std::vector<Object> Snake::getBody()
{
	return body;
}

Direction Snake::getDirection()
{
	return direction;
}

void Snake::CreateBody(int x, int y)
{
	switch (direction)
	{
	case Direction::UP:
		body.push_back(Object(x, y - 1));
		break;

	case Direction::DOWN:
		body.push_back(Object(x, y + 1));
		break;

	case Direction::LEFT:
		body.push_back(Object(x - 1, y));
		break;

	case Direction::RIGHT:
		body.push_back(Object(x + 1, y));
		break;
	}
}

void Snake::MoveBody()
{
	if (body.size() < 1) return;
	Object temp = body[0];
	Object temp2;
	body[0].setX(prevX);
	body[0].setY(prevY);
	for (int i = 1; i < body.size(); i++)
	{
		temp2 = body[i];
		body[i] = temp;
		temp = temp2;
	}
}

void Snake::Grow()
{
	if (body.size() == 0)
		CreateBody(getX(), getY());
	else
		CreateBody(body.back().getX(), body.back().getY());
}

void Snake::Move(Direction direction)
{
	switch (direction)
	{
	case Direction::UP:
		if (this->direction != Direction::DOWN)
		{
			prevX = getX();
			prevY = getY();
			MoveUp();
			this->direction = direction;
		}
		break;

	
	case Direction::DOWN:
		if (this->direction != Direction::UP)
		{
			prevX = getX();
			prevY = getY();
			MoveDown();
			this->direction = direction;
		}
		break;

	case Direction::LEFT:
		if (this->direction != Direction::RIGHT)
		{
			prevX = getX();
			prevY = getY();
			MoveLeft();
			this->direction = direction;
		}
		break;

	case Direction::RIGHT:
		if (this->direction != Direction::LEFT)
		{
			prevX = getX();
			prevY = getY();
			MoveRight();
			this->direction = direction;
		}
		break;
	}
}

bool Snake::isCollideSelf()
{
	for (int i = 0; i < body.size(); i++)
	{
		if (getX() == body[i].getX() && getY() == body[i].getY()) return true;
	}
	return false;
}

File: PySnake/classes/snake.py
from Snake.classes.direction import Direction
from Snake.classes.object import Object


class Snake(Object):
    prev_x = 0
    prev_y = 0
    direction = Direction.RIGHT
    body = []
    
    def createBody(self, x, y):
        self.body.append(Object(x, y, self.sprite))
            

    def grow(self):
        if not self.body:
            Snake.createBody(self, self.x, self.y)
        else:
            Snake.createBody(self, self.body[-1].x, self.body[-1].y)

    def moveBody(self):
        if (not self.body): return
        temp1_x = self.body[0].x
        temp1_y = self.body[0].y
        self.body[0].x = self.prev_x
        self.body[0].y = self.prev_y
        for i in range(1, len(self.body)):
            temp2_x = self.body[i].x
            temp2_y = self.body[i].y
            self.body[i].x = temp1_x
            self.body[i].y = temp1_y
            temp1_x = temp2_x
            temp1_y = temp2_y

    def move(self, direction):

        if direction == Direction.UP:
            self.direction = Direction.UP
        elif direction == Direction.DOWN:
            self.direction = Direction.DOWN
        elif direction == Direction.LEFT:
            self.direction = Direction.LEFT
        else:
            self.direction = Direction.RIGHT

    def isCollideSelf(self):
        for body in self.body:
            if self.x == body.x and self.y == body.y:
                return True
        return False

    def moveDirection(self):
        speed = 32
        x = 0
        y = 0
        if self.direction == Direction.UP:
            y = -speed
        elif self.direction == Direction.DOWN:
            y = speed
        elif self.direction == Direction.LEFT:
            x = -speed
        else:
            x = speed

        self.prev_x = self.x
        self.prev_y = self.y
        self.x += x
        self.y += y